src/goto-instrument/contracts/contracts.cpp

```diff
@@ -235,6 +271,54 @@ void code_contractst::check_apply_loop_contracts(
     loop_end->set_condition(boolean_negate(loop_end->get_condition()));
 }
 
+// Create a lexicographic less-than relation between two tuples of variables.
+// This is used in the implementation of multidimensional decreases clauses.
+static exprt create_lexicographic_less_than(
+  const std::vector<symbol_exprt> &lhs,
+  const std::vector<symbol_exprt> &rhs)
+{
+  PRECONDITION(lhs.size() == rhs.size());
+
+  if(lhs.empty())
+  {
+    return false_exprt();
+  }
+
+  // Store conjunctions of equalities.
+  // For example, suppose that the two input vectors are <s1, s2, s3> and <l1,
+  // l2, l3>.
+  // Then this vector stores <s1 == l1, s1 == l1 && s2 == l2,
+  // s1 == l1 && s2 == l2 && s3 == l3>.
+  // In fact, the last element is unnecessary, so we do not create it.
+  exprt::operandst equality_conjunctions(lhs.size());
+  equality_conjunctions[0] = binary_relation_exprt(lhs[0], ID_equal, rhs[0]);
+  for(unsigned int i = 1; i < equality_conjunctions.size() - 1; i++)
+  {
+    exprt current_component_equality =
+      binary_relation_exprt(lhs[i], ID_equal, rhs[i]);
+    equality_conjunctions[i] =
+      and_exprt(equality_conjunctions[i - 1], current_component_equality);
+  }
+
+  // Store inequalities between the i-th components of the input vectors
+  // (i.e. lhs and rhs).
+  // For example, suppose that the two input vectors are <s1, s2, s3> and <l1,
+  // l2, l3>.
+  // Then this vector stores <s1 < l1, s1 == l1 && s2 < l2, s1 == l1 &&
+  // s2 == l2 && s3 < l3>.
+  exprt::operandst lexicographic_individual_comparisons(lhs.size());
+  lexicographic_individual_comparisons[0] =
+    binary_relation_exprt(lhs[0], ID_lt, rhs[0]);
+  for(unsigned int i = 1; i < lexicographic_individual_comparisons.size(); i++)
+  {
+    exprt current_component_less_than =
+      binary_relation_exprt(lhs[i], ID_lt, rhs[i]);
```

Comment:
Consider removing redundant type annotation (`exprt`):
```suggestion
    binary_relation_exprt component_i_less_than{lhs[i], ID_lt, rhs[i]};
```